### PR TITLE
Electrolyzer Constant Efficiency 

### PR DESCRIPTION
**Pull request by AlexMGTNO**
Formulation of the constant efficiency constraints for the electrolyzers + test suite

### CODE REVIEW

File: tests/test_electrolyzer.py [hunk 1]
```diff
@@ -187,11 +175,154 @@ def energy_system_options(self):
...
+        This test is to check the functioning the example with an offshore wind farm in combination
+        with an electrolyzer and hydrogen storage. The electrolyzer is modelled as the option
+        CONSTANT_EFFICIENCY.
+
+        Checks:
+        - The objective value with the revenue included
+        - Check the bounds on the electrolyzer
+        - Check the setpoint for the windfarm
+        - Check the max production profile of the windfarm
+        - Check the electrolyzer inequality constraints formulation
```
Comment (FJanssen-TNO) [inline tests/test_electrolyzer.py:189]: constant efficiency constraint.

File: tests/test_electrolyzer.py [hunk 2]
```diff
@@ -187,11 +175,154 @@ def energy_system_options(self):
...
+        """
+        This test is to check the functioning the example with an offshore wind farm in combination
+        with an electrolyzer and hydrogen storage. The electrolyzer is modelled as the option
+        CONSTANT_EFFICIENCY.
+
+        Checks:
+        - The objective value with the revenue included
+        - Check the bounds on the electrolyzer
+        - Check the setpoint for the windfarm
+        - Check the max production profile of the windfarm
```
Comment (FJanssen-TNO) [inline tests/test_electrolyzer.py:188]: I don't think it is necessary to check all of these values again, the main focus of this test is to check that the electrolyzer efficiency is constant and that is calculated correctly.

File: tests/test_electrolyzer.py [hunk 3]
```diff
@@ -187,11 +175,154 @@ def energy_system_options(self):
...
+
+        """
+        import models.unit_cases_electricity.electrolyzer.src.example as example
+        from models.unit_cases_electricity.electrolyzer.src.example import (
+            MILPProblemConstantEfficiency,
+        )
+
+        base_folder = Path(example.__file__).resolve().parent.parent
+
+        solution = run_esdl_mesido_optimization(
```
Comment (FJanssen-TNO) [inline tests/test_electrolyzer.py:199]: Think about ensuring that in the test case, the electrolyzer will not be used, e.g. when max power to electrolyzr would be smaller than minimum load (factor 100 I believe).
Change the profile to ensure it and check on it, or change the value of theprofile in the problem definition.

File: .gitignore [hunk 4]
```diff
@@ -35,3 +35,5 @@ run_pipe_compare.py
 *_GrowOptimized.esdl 
 *_GrowOptimized_esdl_string.esdl 
 *_Simulation.esdl
+# Alejandro Martin
+requirements.txt
```
[the remark this answers is not in the document]
Comment (AlexMGTNO) [reply .gitignore:39]: Changed to requirements_amg_.txt

File: src/mesido/electricity_physics_mixin.py [hunk 5]
```diff
@@ -611,52 +611,87 @@ def __electrolyzer_path_constaint(self, ensemble_member):
...
-            constraints.extend(
-                [
-                    (
+            if options["electrolyzer_efficiency"] == ElectrolyzerOption.CONSTANT_EFFICIENCY:
+                nominal = (
+                    self.variable_nominal(f"{asset}.Gas_mass_flow_out")
+                    * self.variable_nominal(f"{asset}.Power_consumed")
+                ) ** 0.5
+                var_name = self.__asset_is_switched_on_map[asset]
+                asset_is_switched_on = self.state(var_name)
```
[the remark this answers is not in the document]
Comment (AlexMGTNO) [reply src/mesido/electricity_physics_mixin.py:626]: Moved to the beginning of the if-else statement

File: src/mesido/electricity_physics_mixin.py [hunk 6]
```diff
@@ -611,52 +611,87 @@ def __electrolyzer_path_constaint(self, ensemble_member):
...
+                            -np.inf,
+                            0.0,
+                        ),
+                    ]
+                )
+            elif (
+                options["electrolyzer_efficiency"]
+                == ElectrolyzerOption.LINEARIZED_THREE_LINES_EQUALITY
+            ):
+                pass
```
[the remark this answers is not in the document]
Comment (AlexMGTNO) [reply src/mesido/electricity_physics_mixin.py:693]: Added

File: tests/test_electrolyzer.py (hunk 2 above)
Comment (AlexMGTNO) [reply tests/test_electrolyzer.py:188]: Reduced test set to just contain the constant efficiency check

File: tests/test_electrolyzer.py (hunk 1 above)
Comment (AlexMGTNO) [reply tests/test_electrolyzer.py:189]: Changed

File: tests/test_electrolyzer.py (hunk 3 above)
Comment (AlexMGTNO) [reply tests/test_electrolyzer.py:195]: Changed the wind profile to 49MW, as the minimum load of the electrolyzer is 50MW

File: .gitlab-ci.yml [hunk 7]
```diff
@@ -51,7 +51,7 @@ pages:
   script:
     - pip install -e .
     - cd docs
-    - pip install -r requirements.txt
+    - pip install -r requirements_amg_.txt
```
Comment (FJanssen-TNO) [inline .gitlab-ci.yml:54]: pleas change this back to requirements.txt

File: .readthedocs.yaml [hunk 8]
```diff
@@ -20,4 +20,4 @@ sphinx:
 python:
   install:
   - path: .
-  - requirements: docs/requirements.txt
+  - requirements: docs/requirements_amg_.txt
```
Comment (FJanssen-TNO) [inline .readthedocs.yaml:23]: requirements.txt

File: tests/test_electrolyzer.py (hunk 3 above)
Comment (FJanssen-TNO) [reply tests/test_electrolyzer.py:195]: Ideally the profile is normally above 50MW, but only for one timestep below 50MW, then you are sure that there is not another reason why the electrolyzer is turned off.

File: tests/test_electrolyzer.py (hunk 2 above)
Comment (FJanssen-TNO) [reply tests/test_electrolyzer.py]: Yes that is perfect, one small one to add, is a check on the electricity power conservation, this ensures that the answer is also always reasonable. If you add a TODO for this that is fine, as I have generalized part of these checks which are in a PR for this week.

File: tests/test_electrolyzer.py (hunk 3 above)
Comment (FJanssen-TNO) [reply tests/test_electrolyzer.py:195]: Ideally the profile is normally above 50MW, but only for one timestep (not the first) below 50MW, then you are sure that there is not another reason why the electrolyzer is turned off.

**Review by FJanssen-TNO — APPROVED** [review]
Looks good! If it passes the pipeline we can squash and merge.

**Pull request merged**